File: src/utils/openai.js
// ここにopenaiの機能を実装する

console.log("Hello, OpenAI!");

// 以下のコマンドで実行
// cd .\src\utils\
// node .\openai.ts

// OpenAIからテキストのレスポンスが返ってくる
// 画像を送ってレスポンスが返ってくる

import base64 from "base64-js";
import fs from "fs";
import OpenAI from "openai";
import path from "path";
import { fileURLToPath } from "url";

const openai = new OpenAI({
  apiKey: "HogeFuga"
});

function encodeImage(imagePath) {
  const imageBuffer = fs.readFileSync(imagePath);
  return base64.fromByteArray(imageBuffer);
}

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// 画像のパスを正しく設定
const imagePath = path.resolve(__dirname, "./image.png");

const base64Image = encodeImage(imagePath);

const chatCompletion = await openai.chat.completions.create({
    model: "gpt-4o-mini",
    messages: [
      {
        role: "user",
        content: [
          {
            type: "text",
            text: "これは鳥の画像ですか?Yes:1,No:0で数字だけ返して"
          },
          {
            type: "image_url",
            // ここで image_url をオブジェクト形式に修正
            image_url: {
              url: `data:image/png;base64,${base64Image}` // 画像データURLをオブジェクトの url プロパティとして指定
            }
          }
        ]
      }
    ]
});

console.log(chatCompletion.choices[0].message.content);

// res.status(200).json({
//     state: "ok",
//     data: chatCompletion
// });

File: src/utils/OpenAI.py
# pythonのサンプル
import base64
import requests
import time
from dotenv import load_dotenv
import os
import json

load_dotenv(dotenv_path='.env.local')

api_key = os.getenv("OPENAI_API_KEY")

def encode_image(image_path):
    with open(image_path, "rb") as image_file:
        return base64.b64encode(image_file.read()).decode('utf-8')

image_path = "js.png"

base64_image = encode_image(image_path)

headers = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {api_key}"
}

payload = {
    "model": "gpt-4o-mini",
    "messages": [
        {
            "role": "user",
            "content": [
                {
                    "type": "text",
                    # 肉, 野菜, 魚介類, 麺, 揚げ物, 穀物, パン, デザート・スイーツ
                    "text": "この画像はNext.jsのプログラムを書いている様子ですか?",
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": f"data:image/jpeg;base64,{base64_image}",
                        "detail": "auto",
                    }
                }
            ]
        }
    ],
    "max_tokens": 100,
}

start_time = time.time()
response = requests.post("https://api.openai.com/v1/chat/completions", headers=headers, json=payload)
end_time = time.time()


response_data = response.json()
print(response_data)
# content = response_data['choices'][0]['message']['content']
# json_content = content.strip("```json\n").strip("\n```")
# parsed_json = json.loads(json_content)
# for key, value in parsed_json.items():
#     print(f"{key}: {value}%")

print(f"\nResponse time: {end_time - start_time} seconds")
